### ISSUE TITLE
Error in vendor/magento/module-shipping/Model/Config/Source/Allmethods.php - public function toOptionArray

### ISSUE THREAD

**Issue by Tantuss** [magento/magento2#13136]
### Preconditions

Magento 2.2.2 but also 2.2.1

### Steps to reproduce

This code is used by an external extension, but this should just work.
See also: https://github.com/magmodules/magento2-channable/issues/28
But this is actually not the extensions bug, it is a bug in Magento 2.2 code.
See below my explanation.
I have no idea where else this is used in Magento, but if it is, there should be also an error visible.

### Expected result

Dropdown box with all carriers, active and inactive.

### Actual result

1 exception(s):
Exception #0 (Exception): Notice: Array to string conversion in vendor/magento/module-shipping/Model/Config/Source/Allmethods.php on line 61

Exception #0 (Exception): Notice: Array to string conversion in vendor/magento/module-shipping/Model/Config/Source/Allmethods.php on line 61
#0 vendor/magento/module-shipping/Model/Config/Source/Allmethods.php(61): Magento\Framework\App\ErrorHandler->handler(8, 'Array to string...', '/Users/jw/Proje...', 61, Array)
#1 vendor/magento/module-config/Model/Config/Structure/Element/Field.php(458): Magento\Shipping\Model\Config\Source\Allmethods->toOptionArray(false)
#2 vendor/magento/module-config/Model/Config/Structure/Element/Field.php(376): Magento\Config\Model\Config\Structure\Element\Field->_getOptionsFromSourceModel(Object(Magento\Shipping\Model\Config\Source\Allmethods))

----
Looking into this Magento code on this line:
'label' => '[' . $carrierCode . '] ' . $methodTitle,

At the $carrierCode = 'usps', I get an array returned.
This results in the error above.

So a simple quick fix is to change:

$methods[$carrierCode]['value'][] = [
          'value' => $carrierCode . '_' . $methodCode,
              'label' => '[' . $carrierCode . '] ' . $methodTitle,
];

into:
if(!is_array($methodTitle))
                    $methods[$carrierCode]['value'][] = [
                        'value' => $carrierCode . '_' . $methodCode,
                        'label' => '[' . $carrierCode . '] ' . $methodTitle,
                    ];

But this is not the complete correct solution, but it makes my page working.

**UPD** from @joni-jones

To reproduce this issue you need to manually insert `carries/usps/allowed_methods` `NULL` into `core_config_data` table and call `\Magento\Shipping\Model\Config\Source\Allmethods::toOptionArray` method.

The issue not in the `\Magento\Shipping\Model\Config\Source\Allmethods::toOptionArray` but how `\Magento\Usps\Model\Carrier::getAllowedMethods` (and other similar methods) processes invalid configuration.

**Comment by Tantuss** [comment]
To make it more complete..
If you look here:
vendor/magento/module-usps/Model/Carrier.php

Line 1148: public function getAllowedMethods()

You see that the return of this function is an array.

So the error is a logical result.

**Comment by YevSent** [comment]
Hi, @Tantuss, did you get the same error on the fresh installation of Magento without extensions? I've checked described case and I didn't get any errors.

According to a described example with `USPS`, the [$carrierMethods](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Shipping/Model/Config/Source/Allmethods.php#L49) contains array in the following format:
```
array (
  '0_FCLE' => 
  Magento\Framework\Phrase::__set_state(array(
     'text' => 'First-Class Mail Large Envelope',
  )),
...
  1 => 
  Magento\Framework\Phrase::__set_state(array(
     'text' => 'Priority Mail',
  )),
...
  'INT_20' => 
  Magento\Framework\Phrase::__set_state(array(
     'text' => 'Priority Mail International Small Flat Rate Envelope',
  )),
)
```
So, [$methodTitle](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Shipping/Model/Config/Source/Allmethods.php#L58) is a [string](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Usps/Model/Carrier.php#L1153), not an array.

**Comment by Tantuss** [comment]
Of course I do not try it on a fresh installation without any configurations.
I assume Magento is not developed to be only installed default and that configuration should be done.

Set in the core_config_data table the path:
carriers/usps/allowed_methods
to a value of 'NULL'.

(so basically nothing selected)

Then you get back the array at the wrong variable.

So probably the mistake is in the USPS extension. So here:
vendor/magento/module-usps/Model/Carrier.php
Line 1148: public function getAllowedMethods()

**Comment by YevSent** [comment]
> Of course I do not try it on a fresh installation without any configurations.
I assume Magento is not developed to be only installed default.

It helps to understand if extensions could be a reason for this issue.

> Set in the core_config_data table the path:
carriers/usps/allowed_methods
to a value of 'NULL'.
(so basically nothing selected)

This behavior doesn't seem correct because you don't have a possibility to don't select at least one allowed method, also, by default, the `Use system value` option is enabled for allowed shipping methods and the configuration will be read from XML because there are no any USPS configurations in `core_config_data` table.

So, the described case is artificial it might happen only if you manually set invalid data into the database. But agree, the code should validate such behavior.

**Comment by Tantuss** [comment]
It might be possible in Magento 1 and it might be there because of migration from Magento 1 to 2.

**Comment by Tantuss** [comment]
> This behavior doesn't seem correct because you don't have a possibility to don't select at least one allowed method, also, by default, the Use system value option is enabled for allowed shipping methods and the configuration will be read from XML because there are no any USPS configurations in core_config_data table.

It might seem not correct, but it is definitely possible to do it from the user interface. Just set 1 as active, saved, checked db and it was there. Then unselected that 1 and there was again NULL in the database.

So it is not only related to migration, it is also just easy possible from the UI.

**Comment by Waldemar-Denke** [comment]
I have same error. Magento 2.2.2, IWD_Opc Extension, Migration from M1 to M2.
On the configuration page of IWD_Opc in magento backend i get the error.
Figured out, that DHL, UPS and Fedex are objects. USPS is an array. 

To figure out the problem i've changed code of `toOptionArray()`:

```
{
        $methods = [['value' => '', 'label' => '']];
        $carriers = $this->_shippingConfig->getAllCarriers();
        foreach ($carriers as $carrierCode => $carrierModel) {
            if (!$carrierModel->isActive() && (bool)$isActiveOnlyFlag === true) {
                continue;
            }
            $carrierMethods = $carrierModel->getAllowedMethods();
            if (!$carrierMethods) {
                continue;
            }
            $carrierTitle = $this->_scopeConfig->getValue(
                'carriers/' . $carrierCode . '/title',
                \Magento\Store\Model\ScopeInterface::SCOPE_STORE
            );
            $methods[$carrierCode] = ['label' => $carrierTitle, 'value' => []];
            foreach ($carrierMethods as $methodCode => $methodTitle) {
                if(is_string($methodTitle)){
                    $methods[$carrierCode]['value'][] = [
                        'value' => $carrierCode . '_' . $methodCode,
                        'label' => '[' . $carrierCode . '] ' . $methodTitle,
                    ];
                } else if (is_object($methodTitle) ){
                    $methods[$carrierCode]['value'][] = [
                        'value' => $carrierCode . '_' . $methodCode,
                        'label' => '[' . $carrierCode . '] ' . $methodTitle->getText(),
                    ];
                } else if(is_array($methodTitle)) {
                    foreach($methodTitle as $key => $value){
                        $methods[$carrierCode]['value'][] = [
                            'value' => $carrierCode . '_' . $key,
                            'label' => '[' . $carrierCode . '] ' . $value->getText(),
                        ];
                    }
                }
            }
        }
        return $methods;
    }
```
possible fix is to check if the method is active or not, directly in the module.

**Comment by magento-engcom-team** [comment]
@Tantuss, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by magento-engcom-team** [comment]
Hi @diazwatson. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by orozcodiaz** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @oroskodias. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @oroskodias, here is your [Magento instance](https://i-13136-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-13136-2-3-develop.instances.magento-community.engineering/admin](https://i-13136-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by VladimirZaets** [comment]
Hi @Tantuss. Thank you for your report. 
The issue has been fixed in magento/magento2#25315 by @mrodespin in 2.4-develop branch
Related commit(s):
 - [529c46f8d445cfb27c58519bf0caab398c15d1c7](https://github.com/magento/magento2/commit/529c46f8d445cfb27c58519bf0caab398c15d1c7)
 - [c4a3c0084eeb076f3620f98b716cd04e8113b004](https://github.com/magento/magento2/commit/c4a3c0084eeb076f3620f98b716cd04e8113b004)
 - [fd7633b28417b6e621808ae9c10cbb707f1b7821](https://github.com/magento/magento2/commit/fd7633b28417b6e621808ae9c10cbb707f1b7821)

The fix will be available with the upcoming 2.4.0 release.
